### ISSUE TITLE
JSON output is locale-dependent and is not cross-platform.

### ISSUE THREAD

**Issue by sergey-shambir** [dropbox/json11#38]
json11 uses locale-dependent functions for conversion between double and it's string representation. On Ubuntu with Russian language it prints doubles with comma like '3,1415926'. On Windows with Russian language pack or any OS with English language it prints '3.1415926' and parser also expects '3.1415926' so '3,1415926' cannot be parsed.

**Comment by peterritter** [comment]
Sergey

Thanks for pointing that out. I'm just a user myself, not any of the
developers, and I hadn't  thought about this issue so far. This is actually
a problem for me too. I wonder if the json specification mandates how
floating point numbers are to presented and whether json is cross platform
by definition.

Peter

**Comment by crazy2be** [comment]
It does: http://www.json.org/

It is very clear that a period is supposed to be used for the decimal point if you look at the parsing diagrams.

**Comment by j4cbo** [comment]
This is definitely a bug. We likely don't have time to implement a proper fix, but I'd love to see a pull request. Options include:
- Use snprintf_l if available or emulate somehow if not
- Loop through the string after conversion and replace `localeconv()->decimal_point` with `.`
- Embed our own copy of `grisu2` (I think this is my preferred option)

**Comment by Dushistov** [comment]
`json11` also have str->double bug, yes, in `parse_number()` it checks that number have "." as delimiter, but after that it uses `std::strtod`, and this function depend on locale, so `10.5` it can convert to `10.0`, because of it expects another delimiter.

**Comment by Dushistov** [comment]
What about import code from `rapid json`, it has very fast string<->double routies: https://github.com/miloyip/rapidjson/tree/master/include/rapidjson/internal

and the code is under MIT license?

**Comment by Dushistov** [comment]
Also there is https://github.com/google/double-conversion from author of `girus3` (BSD like license), what about use it as submodule?

**Comment by Dushistov** [comment]
I convert json11 to usage of double-conversion (for string->double and double->string): https://github.com/Dushistov/json11/commit/115fc7609ddd8bb2661ea0dd5ce54a5b347c0c62

nativejson-benchmark from here
https://github.com/miloyip/nativejson-benchmark give:

Before:
Parse Time (ms) 51	
Stringify Time (ms) 109

After:
Parse Time (ms) 41
Stringify Time (ms) 23

As you can see speedup 24% for parsing, and 374% for dump,
plus it not depend on locale.

What do you think?

**Comment by artwyman** [comment]
I like the idea of fixing this, and like the sound of those benchmarks.  A submodule seems quite heavy weight for json11, which is currently a very tiny library which is easy to integrate.  Dropbox's own build environment wouldn't have any trouble with a submodule, but it's a lot of overhead for other users.  Is there a simpler option?

I'll get an answer on whether the license is okay, but I suspect it would be.

**Comment by artwyman** [comment]
Pull requests welcome.

**Comment by skypjack** [comment]
@artwyman 

@j4cbo suggested the followings:

- `snprintf_l` if available or emulate somehow if not
- Loop through the string after conversion and replace `localeconv()->decimal_point` with `.`
- Embed our own copy of _grisu2_ (I think this is my preferred option)

Are all good for you? Probably point 2 would be a quick (though pretty dirty) solution while waiting for someone to implement or port _grisu2_.

**Comment by artwyman** [comment]
I don't have time to go deep, but they all sound plausible.  My primary concerns on any new dependencies would be how portable they are, particularly to the mobile platforms where Dropbox uses json11.  For something hand-tweaked like #2 I'd just want to make sure it has minimal risk for unintended impact (i.e. don't accidentally mess with commas in strings) and minimal performance impact.

**Comment by artwyman** [comment]
Thanks for taking a stab at it.  Please phrase your submissions in the form of a Pull Request, though, not a patch.

On a quick look, I wonder about the potential performance difference between stringstream and snprintf, and whether that could add up when printing a lot of numbers.  I wouldn't know for sure if that's an issue without measuring, though.

**Comment by ashaduri** [comment]
C++17 has std::from_chars() and std::to_chars() specifically for this kind of usage (JSON, etc...). They're also supposed to be very fast (faster than the usual locale-aware functions).
An ifdef to enable these for people using C++17 would be very nice.

http://en.cppreference.com/w/cpp/utility/from_chars
http://en.cppreference.com/w/cpp/utility/to_chars

**Comment by alekseyt** [comment]
Nah, i'm good. I can only note that it's too trivial to be copyrighted, so anyone are welcome to do whatever they want with those patches.
